core/src/labels/labels.cpp

```diff
@@ -169,187 +174,135 @@ void Labels::skipTransitions(const std::vector<std::unique_ptr<Style>>& _styles,
     }
 }
 
-void Labels::checkRepeatGroups(std::vector<Label*>& _visibleSet) const {
+bool Labels::labelComparator(const LabelEntry& _a, const LabelEntry& _b) {
+    if (_a.proxy != _b.proxy) {
+        return _b.proxy;
+    }
+
+    if (_a.priority != _b.priority) {
+        return _a.priority < _b.priority;
+    }
+
+    auto l1 = _a.label;
+    auto l2 = _b.label;
+
+    // Note: This causes non-deterministic placement, i.e. depending on
+    // navigation history.
+    if (l1->occludedLastFrame() != l2->occludedLastFrame()) {
+        return l2->occludedLastFrame();
+    }
+
+    // if (l1->options().repeatGroup != l2->options().repeatGroup) {
+    //     return l1->options().repeatGroup < l2->options().repeatGroup;
+    // }
+
+    if (l1->type() == Label::Type::line && l2->type() == Label::Type::line) {
+        // Prefer the label with longer line segment as it has a chance
+        return glm::length(l1->transform().modelPosition1 - l1->transform().modelPosition2) >
+               glm::length(l2->transform().modelPosition1 - l2->transform().modelPosition2);
+    }
+
+    if (l1->hash() != l2->hash()) {
+        return l1->hash() < l2->hash();
+    }
+
+    return l1 < l2;
+}
 
-    struct GroupElement {
-        Label* label;
+void Labels::sortLabels() {
```

Comment:
yes, it can be inlined again. I've split it up to see the time spent for the different phases more easily in perf